ibis/sql/postgres/tests/test_udf.py

```diff
@@ -0,0 +1,173 @@
+"""Test support for already-defined UDFs in Postgres"""
+
+import random
+import string
+
+import pytest
+
+import pandas as pd
+import ibis.expr.datatypes
+import ibis.sql.postgres.udf.api
+from ibis.sql.postgres.client import PostgreSQLClient
+
+
+datatypes = ibis.expr.datatypes
+
+# mark test module as postgresql (for ability to easily exclude,
+# e.g. in conda build tests)
+# (Temporarily adding `postgis` marker so Azure Windows pipeline will exclude
+#     pl/python tests.
+#     TODO: update Windows pipeline to exclude postgres_extensions
+#     TODO: remove postgis marker below once Windows pipeline updated
+pytestmark = [
+    pytest.mark.postgresql,
+    pytest.mark.postgis,
+    pytest.mark.postgres_extensions,
+]
+
+# Database setup (tables and UDFs)
+
+
+test_schema = None
+
+
+def gen_schema_name(basename='test'):
+    """Generate a random alpha string starting with 'test_' to be used as a
+    test schema name"""
+    schema_name = '{}_{}'.format(
+        basename,
+        ''.join([random.choice(string.ascii_lowercase) for i in range(6)])
+    )
+    return schema_name
+
+
+def create_test_schema(connection: PostgreSQLClient, max_tries=20):
+    """Create a random test schema, necessary to allow for parallel pytest
+    testing"""
+    schema_exists = True
+    tries = 0
+    while schema_exists and tries < max_tries:
+        new_name = gen_schema_name()
+        sql_check_schema = """SELECT count(schema_name) as n_recs
+        FROM information_schema.schemata
+        WHERE lower(schema_name) = lower('{}');""".format(new_name)
+        df_result = pd.read_sql(sql_check_schema, connection.con)
+        schema_exists = df_result['n_recs'].iloc[0] > 0
+        tries += 1
+    assert tries > 0
+    global test_schema
```

Comment:
Why did you decide to make this a global variable as opposed to a fixture or a function that returns the schema?

Reply:
Do fixtures re-execute every time they are used, or are they cached? Because I don't want it to try re-creating the test_schema every time the fixture is used.

Reply:
You can specify the scope at which they run, the "highest" scope being `'session'` which runs exactly once per test session. You can specify this using `@pytest.fixture(scope='session')`.

We have a number of tests that perform stateful manipulations (like creating a table) on databases (Impala in particular) and we're able to run those in parallel, so in theory we should be able to do the same with postgres.

Reply:
Fixed. No longer using a global and doing it as a fixture with session scope